### ISSUE TITLE
NullPointer with bad cluster node list

### ISSUE THREAD

**Issue by taer** [mp911de/lettuce#173]
With the following output from a cluster list, I get a NPE from Lettuce. 
```
f1e2d6bcb939c5e97cdf1372bde3958620 10.37.110.63:7000 slave 2a83151d8b5fef27351623a1ed64112f72b5692f 0 1452553664848 43 connected
52b2590d15237fe51b1887804f679fbd10fcb1d8 10.37.110.68:7000 slave 1f98e0e56e115375a7369f5daf38b607960a771d 0 1452553664346 45 connected
e55f975d1cec20e46680451088ce0d8d07b1115d :0 slave,fail,noaddr d1cd1db25e5198e132a27d527cf65a6c872dcd5c 1449160058028 1449160053146 46 disconnected
2a83151d8b5fef27351623a1ed64112f72b5692f 10.37.110.69:7000 master - 0 1452553662842 43 connected 3829-6787 7997-9999
66c43dcdfccac59dbb2afd3455eecc829370bbb1 10.37.110.62:7000 slave 2a83151d8b5fef27351623a1ed64112f72b5692f 0 1452553663844 43 connected
d1cd1db25e5198e132a27d527cf65a6c872dcd5c 10.37.110.70:7000 myself,master - 0 0 46 connected 10039-14999
1f98e0e56e115375a7369f5daf38b607960a771d 10.37.110.65:7000 master - 0 1452553663844 45 connected 0-3828 6788-7996 10000-10038 15000-16383
```

```
! java.lang.NullPointerException: null
! at com.lambdaworks.redis.cluster.PooledClusterConnectionProvider.getPartition(PooledClusterConnectionProvider.java:235) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.redis.cluster.PooledClusterConnectionProvider.access$700(PooledClusterConnectionProvider.java:45) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.redis.cluster.PooledClusterConnectionProvider$ConnectionFactory.load(PooledClusterConnectionProvider.java:512) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.redis.cluster.PooledClusterConnectionProvider$ConnectionFactory.load(PooledClusterConnectionProvider.java:483) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3527) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2319) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2282) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2197) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.com.google.common.cache.LocalCache.get(LocalCache.java:3937) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
! at com.lambdaworks.redis.cluster.PooledClusterConnectionProvider.getReadFromConnections(PooledClusterConnectionProvider.java:170) ~[tank-1.1.1-SNAPSHOT.jar:1.1.1-SNAPSHOT]
```

The constructor of the RedisClusterNode is at fault. He allows a null uri via constructor, but disallows via setter. It's constructed from ClusterPartitionParser.parseNode(). If that gets a line like with an ip of :0, it skips the uri setting, which then later NPEs with the above trace.

In this case, this is a failed node that was removed from the cluster, but still lives around on some nodes.

I think the ClusterPartitionParser should key off the noaddr flag. Perhaps returning null from the parseNode function, which would then NOT add to the partition list. 

Or keep it, but make the consumer code in PooledClusterConnectionProvider.getPartitions() skip non-active nodes.  RedisClusterNode.getURI() is public though, and seems to be used in many places. Having a null uri might mess with other places.

**Comment by taer** [comment]
BTW, this is for 4.0.2.

Working on a patch for the first option. Also, the pom.xml shows 1.6 as the target version. The 3->4 upgrade shows that only java 8 is supported. Are there plans to update to that soon?

Thanks

**Comment by taer** [comment]
#174 Submitted a PR. But I did it against master. What is your current branch strategy?

**Comment by mp911de** [comment]
Hi, 
thank's for the patch. I'm maintaining both branches (master with 3.4 and 4.1 versions in development, see also https://github.com/mp911de/lettuce/wiki#3x-and-4x-branching-policy). I'm about to release 3.4 and 4.1 end of January. 3.x should receive only bugfixes and command updated whereas 4.x becomes the default branch after the 4.1 release.

**Comment by mp911de** [comment]
I'd still allow `RedisClusterNode` without a URI because that's a valid state within Redis (noaddr, `:0` addr). I propose to solve this problem with filtering invalid partitions within the `ClusterTopologyRefresh` not to affect other users of `ClusterPartitionParser`.

After fixing the NPE, you'll still receive an Exception in that case, but it will be a `RedisException` with a message of `Cannot determine a partition to read for slot ...`.

**Comment by taer** [comment]
In that case posted, all slots were covered. In fact, some of the nodes don't know about that node at all. I'll rebuild that, but it's something that can happen in the real world.

For the PR, would you want 2 PRs 3.4 as well as the 4.1 branch?

**Comment by mp911de** [comment]
Right, oversaw this. Would be great to get two PR's. Otherwise I'd merge the master (3.4) PR into 4.1.

**Comment by mp911de** [comment]
Closing this issue as this is resolved for the upcoming 3.4 and 4.1 releases
